components/net/indexeddb/engines/sqlite/serialize.rs

```diff
@@ -0,0 +1,360 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use net_traits::indexeddb_thread::IndexedDBKeyType;
+
+/// A module for encoding and decoding various IndexedDBKeyType types according to the basic strategy:
+///
+/// - Numbers:  0x10 + 8-byte sortable float
+/// - Dates:    0x20 + 8-byte sortable float
+/// - Strings:  0x30 + encoded chars + 0 terminator
+/// - Binaries: 0x40 + encoded octets + 0 terminator
+/// - Arrays:   0x50 + encoded items + 0 terminator
+
+/// Internal helper: convert f64 to sortable u64 per IEEE754.
+fn f64_to_sortable(value: f64) -> u64 {
+    let bits = value.to_bits();
+    if value.is_sign_negative() {
+        bits.wrapping_neg()
+    } else {
+        bits | 0x8000_0000_0000_0000
+    }
+}
+
+/// Internal helper: reverse sortable u64 back to f64.
+fn sortable_to_f64(bits: u64) -> f64 {
+    let orig = if bits & 0x8000_0000_0000_0000 != 0 {
+        bits & 0x7FFF_FFFF_FFFF_FFFF
+    } else {
+        bits.wrapping_neg()
+    };
+    f64::from_bits(orig)
+}
+
+/// Encode a number (f64) with prefix 0x10.
+pub fn encode_number(item: f64) -> Vec<u8> {
+    let mut out = Vec::with_capacity(9);
+    out.push(0x10);
+    out.extend(&f64_to_sortable(item).to_be_bytes());
+    out
+}
+
+/// Decode a number from a byte slice, returning (IndexedDBKeyType, bytes_consumed).
+pub fn decode_number(input: &[u8]) -> Option<(f64, usize)> {
+    if input.get(0)? != &0x10 || input.len() < 9 {
+        return None;
+    }
+    let mut arr = [0u8; 8];
+    arr.copy_from_slice(&input[1..9]);
+    let bits = u64::from_be_bytes(arr);
+    Some((sortable_to_f64(bits), 9))
+}
+
+/// Encode a date with prefix 0x20 and terminator 0x00.
+pub fn encode_date(data: &[u8]) -> Vec<u8> {
```

Comment:
I would find it helpful if we could document the intention of these routines. Are they following some kind of specification, or using another implementation as a model?